packages/patterns/integration/shadow-dom-utils.ts

```diff
@@ -0,0 +1,170 @@
+/**
+ * Utility functions for navigating through shadow DOM boundaries in tests
+ */
+
+type Page = any; // The Page type from integration utils
+
+/**
+ * Traverses through shadow DOM boundaries following a path of selectors
+ * @param page - The page object
+ * @param path - Array of selectors to traverse through shadow DOMs
+ * @returns The element at the end of the path, or null if not found
+ */
+export async function traverseShadowDOM(
+  page: Page,
+  path: string[],
+): Promise<any> {
+  return await page.evaluate((selectors: string[]) => {
+    let current: Document | Element | ShadowRoot | null = document;
+    
+    for (let i = 0; i < selectors.length; i++) {
+      const selector = selectors[i];
+      
+      if (!current) {
+        console.error(`Lost element at step ${i}`);
+        return null;
+      }
+      
+      // Query the current context (document, element, or shadow root)
+      let next: Element | null = null;
+      
+      if (current instanceof Document) {
+        next = current.querySelector(selector);
+      } else if (current instanceof ShadowRoot) {
+        next = current.querySelector(selector);
+      } else if (current instanceof Element) {
+        // First try shadow root if it exists
+        if (current.shadowRoot) {
+          next = current.shadowRoot.querySelector(selector);
+        }
+        // If not found in shadow root (or no shadow root), try regular DOM
+        if (!next) {
+          next = current.querySelector(selector);
+        }
+      }
+      
+      if (!next) {
+        console.error(`Could not find "${selector}" at step ${i}`);
+        return null;
+      }
+      
+      // For the next iteration, we'll search in this element's shadow root if it has one,
+      // otherwise we'll search in the element itself
+      current = next;
+    }
+    
+    // Return the final element (not its shadow root)
+    if (current instanceof ShadowRoot) {
+      return current.host;
+    }
+    return current;
+  }, { args: [path] });
+}
+
+/**
+ * Clicks an element deep in shadow DOM
+ * @param page - The page object
+ * @param path - Array of selectors to traverse through shadow DOMs
+ */
+export async function clickShadowElement(
```

Comment:
the "Shadow" prefix on the functions might be unneeded, the selectors can navigate across shadow boundaries but the resulting elements aren't any different than regular (light dom) elements, and can be thought of as an alternative selector engine

Reply:
May be a notable detail, but in the selector path, descending into shadow roots can only be done by specifying a shadow host as a target in a selector